programs/mango-v4/src/state/token_conditional_swap.rs

```diff
@@ -197,10 +290,74 @@ impl TokenConditionalSwap {
             .to_num()
     }
 
-    pub fn price_in_range(&self, price: f64) -> bool {
+    fn price_in_range(&self, price: f64) -> bool {
         price >= self.price_lower_limit && price <= self.price_upper_limit
     }
 
+    pub fn check_startable(&self, price: f64, now_ts: u64) -> Result<()> {
+        require!(
+            !self.is_expired(now_ts),
+            MangoError::TokenConditionalSwapExpired
+        );
+        require!(
+            self.start_timestamp == 0,
+            MangoError::TokenConditionalSwapAlreadyStarted
+        );
+        match self.tcs_type() {
+            TokenConditionalSwapType::FixedPremium => {
+                require!(
+                    self.price_in_range(price),
+                    MangoError::TokenConditionalSwapPriceNotInRange
+                );
+            }
+            TokenConditionalSwapType::PremiumAuction => {
+                require!(
+                    self.price_in_range(price),
+                    MangoError::TokenConditionalSwapPriceNotInRange
+                );
+            }
+            TokenConditionalSwapType::LinearAuction => {}
+        };
+        Ok(())
+    }
+
+    pub fn is_startable(&self, price: f64, now_ts: u64) -> bool {
+        self.check_startable(price, now_ts).is_ok()
+    }
+
+    pub fn check_triggerable(&self, price: f64, now_ts: u64) -> Result<()> {
+        require!(
+            !self.is_expired(now_ts),
+            MangoError::TokenConditionalSwapExpired
+        );
+        match self.tcs_type() {
+            TokenConditionalSwapType::FixedPremium => {
+                require!(
+                    self.price_in_range(price),
+                    MangoError::TokenConditionalSwapPriceNotInRange
+                );
+            }
+            TokenConditionalSwapType::PremiumAuction => {
+                // Technically the error is lying, since it's also triggerable once started
+                require!(
+                    self.price_in_range(price) || self.passed_start(now_ts),
```

Comment:
why is this check not simply ` self.passed_start(now_ts)`? the `self.price_in_range(price)` would be implicitly taken care once the tcs start ix for premium auction is called

Reply:
I want to allow triggering directly without start ever being called.